---
title: "Snakemake AmpBS-Seq Pipeline"
author: "Maximilian Sch√∂nung"
date: "12/18/2018"
output: html_document
---


<style>
body {
text-align: justify}
</style>

<br>

### AmpBS-Seq workflow

Amplicon bisulfite sequencing (AmpBS-Seq) is a powerful approach for targeted DNA methylation analysis. The present [protocol](workflow.pdf) describes the detailed workflow, reagents required and offers together with this website all the required ressources for a successfull targeted DNA methylation assay.



### Analysis of AmpBS-Seq data

The analysis of AmpBS-Seq data requires several bioinformatic processing steps: <br>

* Adapter Trimming
* Alignment
* Extraction of Methylation Values
* Quality Control


We have developed a [Snakemake Pipeline](https://github.com/HeyLifeHD/amplicon_bisulfite_seq_pipeline) which integrates several publicly available tools into an off-the-shelf pipeline.

The pipeline provides the user with Bismark coverage (.cov) files which can be further explored using this interactive  analysis and quality control pipeline:

<br>

#### **Upload Sample Sheet**
A sample sheet which specifies meta data and annotations for the analyzed samples can be uploaded. The first column of this file corresponds to the coverage file name (ex. "Sample1.cov") and a column "UPN" must be included which assigns each sample a unique identifier. An example sample sheet ("meta.txt") can be downloaded with the sample data.

<br>

#### **Upload Regions**
The analyzed regions must be uploaded as bed-Files.

<br>

#### **Upload .cov Files**
Upload of Bismark coverage Files.

<br>

#### **Coverage Cut-Off**
During the interactive analysis, CpG sites can be filtered based on a minimal number of sequencing reads present at a certain CpG sites.  

